File: plateform/frontend/src/api/maintenance.ts
import type { AssetMaintenanceResponse, IncidentMaintenanceResponse, MaintenanceFilters, UpcomingMaintenanceItem, WorkOrderDetail, WorkOrderListResponse } from "../types/maintenance";
import { apiClient } from "./client";

export function toMaintenanceQuery(filters: MaintenanceFilters): Record<string, string> {
  const query: Record<string, string> = {};
  if (filters.search.trim()) query.search = filters.search.trim();
  if (filters.zone) query.zone = filters.zone;
  if (filters.asset_id.trim()) query.asset_id = filters.asset_id.trim();
  if (filters.asset_type) query.asset_type = filters.asset_type;
  if (filters.maintenance_type) query.maintenance_type = filters.maintenance_type;
  if (filters.priority) query.priority = filters.priority;
  if (filters.status) query.status = filters.status;
  if (filters.overdue) query.overdue = "true";
  if (filters.assigned_to.trim()) query.assigned_to = filters.assigned_to.trim();
  if (filters.sort_by !== "due_at") query.sort_by = filters.sort_by;
  if (filters.sort_order !== "asc") query.sort_order = filters.sort_order;
  return query;
}

export async function fetchWorkOrders(filters: MaintenanceFilters): Promise<WorkOrderListResponse> {
  const { data } = await apiClient.get<WorkOrderListResponse>("/api/maintenance/work-orders", {
    params: toMaintenanceQuery(filters),
  });
  return data;
}

export async function fetchUpcomingMaintenance(): Promise<UpcomingMaintenanceItem[]> {
  const { data } = await apiClient.get<UpcomingMaintenanceItem[]>("/api/maintenance/work-orders/upcoming");
  return data;
}

export async function fetchWorkOrder(workOrderId: string): Promise<WorkOrderDetail> {
  const { data } = await apiClient.get<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(workOrderId)}`,
  );
  return data;
}

export async function fetchAssetMaintenance(assetId: string): Promise<AssetMaintenanceResponse> {
  const { data } = await apiClient.get<AssetMaintenanceResponse>(
    `/api/assets/${encodeURIComponent(assetId)}/maintenance`,
  );
  return data;
}

export async function fetchIncidentMaintenance(incidentId: string): Promise<IncidentMaintenanceResponse> {
  const { data } = await apiClient.get<IncidentMaintenanceResponse>(
    `/api/incidents/${encodeURIComponent(incidentId)}/maintenance`,
  );
  return data;
}

export async function createWorkOrder(payload: Record<string, unknown>): Promise<WorkOrderDetail> {
  const { data } = await apiClient.post<WorkOrderDetail>("/api/maintenance/work-orders", payload);
  return data;
}

export async function assignWorkOrder(id: string, actorName: string, assignedTo: string, note?: string) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/assign`,
    { actor_name: actorName, assigned_to: assignedTo, note },
  );
  return data;
}

export async function rescheduleWorkOrder(id: string, actorName: string, dueAt: string, note?: string) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/reschedule`,
    { actor_name: actorName, due_at: dueAt, note },
  );
  return data;
}

export async function startWorkOrder(
  id: string,
  actorName: string,
  options: { assignedTo?: string; confirmUnassigned?: boolean; note?: string } = {},
) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/start`,
    {
      actor_name: actorName,
      assigned_to: options.assignedTo,
      confirm_unassigned: options.confirmUnassigned ?? false,
      note: options.note,
    },
  );
  return data;
}

export async function addWorkOrderNote(id: string, actorName: string, note: string) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/notes`,
    { actor_name: actorName, note },
  );
  return data;
}

export async function completeWorkOrder(
  id: string,
  actorName: string,
  completionResult: string,
  completionSummary: string,
) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/complete`,
    {
      actor_name: actorName,
      completion_result: completionResult,
      completion_summary: completionSummary,
      confirm: true,
    },
  );
  return data;
}

export async function cancelWorkOrder(id: string, actorName: string, reason: string) {
  const { data } = await apiClient.post<WorkOrderDetail>(
    `/api/maintenance/work-orders/${encodeURIComponent(id)}/cancel`,
    { actor_name: actorName, reason, confirm: true },
  );
  return data;
}

File: plateform/backend/app/api/routes/maintenance.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query

from app.api.deps import get_maintenance_service
from app.schemas.assets import AssetType
from app.schemas.incidents import SortOrder
from app.schemas.maintenance import (
    AssignWorkOrderRequest,
    CancelWorkOrderRequest,
    CompleteWorkOrderRequest,
    CreatePlanRequest,
    CreateWorkOrderRequest,
    DisablePlanRequest,
    MaintenanceNoteRequest,
    MaintenancePlanListResponse,
    MaintenancePlanSummary,
    MaintenancePriority,
    MaintenanceSummary,
    MaintenanceType,
    RescheduleWorkOrderRequest,
    StartWorkOrderRequest,
    UpdatePlanRequest,
    UpcomingMaintenanceItem,
    WorkOrderDetail,
    WorkOrderHistoryResponse,
    WorkOrderListResponse,
    WorkOrderSortField,
    WorkOrderStatus,
)
from app.services.maintenance import MaintenanceService

router = APIRouter(prefix="/maintenance", tags=["maintenance"])


@router.get("/summary", response_model=MaintenanceSummary)
def get_maintenance_summary(
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenanceSummary:
    return service.summary()


@router.get("/work-orders", response_model=WorkOrderListResponse)
def list_work_orders(
    status: WorkOrderStatus | None = Query(default=None),
    priority: MaintenancePriority | None = Query(default=None),
    maintenance_type: MaintenanceType | None = Query(default=None),
    asset_type: AssetType | None = Query(default=None),
    asset_id: str | None = Query(default=None),
    zone: str | None = Query(default=None),
    assigned_to: str | None = Query(default=None),
    overdue: bool | None = Query(default=None),
    due_from: datetime | None = Query(default=None),
    due_to: datetime | None = Query(default=None),
    search: str | None = Query(default=None),
    sort_by: WorkOrderSortField = Query(default=WorkOrderSortField.due_at),
    sort_order: SortOrder = Query(default=SortOrder.asc),
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderListResponse:
    return service.list_work_orders(
        status=status,
        priority=priority,
        maintenance_type=maintenance_type,
        asset_type=asset_type,
        asset_id=asset_id,
        zone=zone,
        assigned_to=assigned_to,
        overdue=overdue,
        due_from=due_from,
        due_to=due_to,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.get("/work-orders/upcoming", response_model=list[UpcomingMaintenanceItem])
def list_upcoming(
    service: MaintenanceService = Depends(get_maintenance_service),
) -> list[UpcomingMaintenanceItem]:
    return service.upcoming()


@router.get("/work-orders/{work_order_id}", response_model=WorkOrderDetail)
def get_work_order(
    work_order_id: str,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.get_work_order(work_order_id)


@router.get("/work-orders/{work_order_id}/history", response_model=WorkOrderHistoryResponse)
def get_work_order_history(
    work_order_id: str,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderHistoryResponse:
    return service.get_history(work_order_id)


@router.post("/work-orders", response_model=WorkOrderDetail)
def create_work_order(
    payload: CreateWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.create_work_order(payload)


@router.post("/work-orders/{work_order_id}/assign", response_model=WorkOrderDetail)
def assign_work_order(
    work_order_id: str,
    payload: AssignWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.assign(work_order_id, payload)


@router.post("/work-orders/{work_order_id}/reschedule", response_model=WorkOrderDetail)
def reschedule_work_order(
    work_order_id: str,
    payload: RescheduleWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.reschedule(work_order_id, payload)


@router.post("/work-orders/{work_order_id}/start", response_model=WorkOrderDetail)
def start_work_order(
    work_order_id: str,
    payload: StartWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.start(work_order_id, payload)


@router.post("/work-orders/{work_order_id}/notes", response_model=WorkOrderDetail)
def add_work_order_note(
    work_order_id: str,
    payload: MaintenanceNoteRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.add_note(work_order_id, payload)


@router.post("/work-orders/{work_order_id}/complete", response_model=WorkOrderDetail)
def complete_work_order(
    work_order_id: str,
    payload: CompleteWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.complete(work_order_id, payload)


@router.post("/work-orders/{work_order_id}/cancel", response_model=WorkOrderDetail)
def cancel_work_order(
    work_order_id: str,
    payload: CancelWorkOrderRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> WorkOrderDetail:
    return service.cancel(work_order_id, payload)


@router.get("/plans", response_model=MaintenancePlanListResponse)
def list_plans(
    asset_id: str | None = Query(default=None),
    enabled: bool | None = Query(default=None),
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenancePlanListResponse:
    return service.list_plans(asset_id=asset_id, enabled=enabled)


@router.get("/plans/{plan_id}", response_model=MaintenancePlanSummary)
def get_plan(
    plan_id: str,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenancePlanSummary:
    return service.get_plan(plan_id)


@router.post("/plans", response_model=MaintenancePlanSummary)
def create_plan(
    payload: CreatePlanRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenancePlanSummary:
    return service.create_plan(payload)


@router.patch("/plans/{plan_id}", response_model=MaintenancePlanSummary)
def update_plan(
    plan_id: str,
    payload: UpdatePlanRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenancePlanSummary:
    return service.update_plan(plan_id, payload)


@router.post("/plans/{plan_id}/disable", response_model=MaintenancePlanSummary)
def disable_plan(
    plan_id: str,
    payload: DisablePlanRequest,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> MaintenancePlanSummary:
    return service.disable_plan(plan_id, payload)
